backend/pcloud/chunkwriter.go

```diff
@@ -0,0 +1,337 @@
+package pcloud
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"sync"
+
+	"github.com/rclone/rclone/backend/pcloud/api"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/lib/rest"
+	"golang.org/x/sync/errgroup"
+)
+
+type chunkWriter struct {
+	client    *rest.Client
+	fs        *Fs
+	chunkSize int64
+	src       fs.ObjectInfo
+	remote    string
+	fd        int64
+	fileID    int64
+
+	mu        sync.RWMutex
+	byteCount int64
+}
+
+func (c *chunkWriter) GetBytesCount() int64 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.byteCount
+}
+
+func (c *chunkWriter) AddBytesCount(count int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.byteCount = c.byteCount + count
+}
+
+func newChunkWriter(
+	ctx context.Context,
+	remote string,
+	src fs.ObjectInfo,
+	srcFs *Fs,
+) (info fs.ChunkWriterInfo, writer *chunkWriter, err error) {
+	client, err := srcFs.newSingleConnClient(ctx)
+	if err != nil {
+		return info, writer, fmt.Errorf("create client: %w", err)
+	}
+	// init an empty file
+	leaf, directoryID, err := srcFs.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return info, writer, fmt.Errorf("resolve src: %w", err)
+	}
+	openResult, err := fileOpenNew(ctx, client, srcFs, directoryID, leaf)
+	if err != nil {
+		return info, writer, fmt.Errorf("open file: %w", err)
+	}
+
+	info = fs.ChunkWriterInfo{
+		ChunkSize:         int64(srcFs.opt.ChunkSize),
+		Concurrency:       srcFs.opt.UploadConcurrency,
+		LeavePartsOnError: true,
+	}
+	writer = &chunkWriter{
+		client:    client,
+		fs:        srcFs,
+		chunkSize: info.ChunkSize,
+		src:       src,
+		remote:    remote,
+		fd:        openResult.FileDescriptor,
+		fileID:    openResult.Fileid,
+	}
+
+	return info, writer, nil
+}
+
+// Abort implements fs.ChunkWriter.
+func (c *chunkWriter) Abort(ctx context.Context) error {
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open aborted: %w", err)
+	}
+	if err := o.Remove(ctx); err != nil {
+		return fmt.Errorf("delete aborted: %w", err)
+	}
+	return nil
+}
+
+// Close implements fs.ChunkWriter.
+func (c *chunkWriter) Close(ctx context.Context) error {
+	// In case the file existed previosly and was larger than the newly written file, we need to truncate it to the right size.
+	byteCount := c.GetBytesCount()
+	fs.Debugf(c.src, "truncate target to %d bytes", byteCount)
```

Comment:
Does this mean that we are now overwriting existing files?

This means that if a transfer fails we've clobbered an existing file - is that correct?

If so then we need to set this feature flag

        PartialUploads           bool // uploaded file can appear incomplete on the fs while it's being uploaded

so rclone knows to upload to a temporary file name and rename it over at the end.

Reply:
@ncw I also tried to get some more details on this:
IMHO PartialUploads would be bad, as we loose the continuation of chunked uploads (see other discussion).
Yet I could make any sorts of handling for existing files, e.g
- delete on failure (current solution)
- [revert on failure](https://docs.pcloud.com/methods/revisions/revertrevision.html)
- leave broken file

In order to have a good solution here, can you tell me what the desired behavior would be on the following cases:

- successful upload (Close is called) of a file that already exists
- failed upload (Abort is called) of a file that already exists
- successful upload (Close is called) of a file that doesn't exist
- failed upload (Abort is called) of a file that doesn't exist

(I deliberately included the obvious cases for sake for completeness)